---
title: "base_Z_score_questions.rmd"
author: "Tim Nieuwenhuis"
date: "6/14/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(patchwork)
source("../../global_in/general_scripts.R")
```

Select items of interest
```{r}
## Tissue R name
tiss <- "muscle_skeletal"
```

load in data
```{r}

#Load genes and profiles
gene_path <- paste0("../../MARCC output/variance_genes_kendall/kendall-",
                    tiss,"-gene-clusters-high-variance.csv")

clust_genes <- read.csv(gene_path)
prof_path <- paste0("../../MARCC output/variance_profiles_kendall/kendall-",
                    tiss,"-cluster-profiles.csv")
clust_profs <- read.csv(prof_path,row.names = 1)

## Load gene clusters

## Load SUBJ data
phen_dat <- read.csv("../../global_in/gtex_phenotypes_v8.csv")
## Load samp data.
samp_dat <- read.delim("../../global_in/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt", fill=T,
                      stringsAsFactors=FALSE, sep="\t", header=T)
## Load common subj_dat
subj_dat <- read.delim("../../global_in/GTEx_Analysis_v8_Annotations_SubjectPhenotypesDS.txt", fill=T,
                      stringsAsFactors=FALSE, sep="\t", header=T)
## Load keys 
key_dat <- read.csv("../../automated_cluster_analysis/data_out/key_table.csv")
```


## Load sample data

Clean and join data


Join_data
```{r}
final_dat <- clust_profs %>% rownames_to_column(var = "SAMPID") %>%
    mutate(SUBJID = GTEx_SAMPID_to_SUBJID(SAMPID)) %>%
    left_join(phen_dat, by = "SUBJID") %>%
    left_join(samp_dat, by = "SAMPID") %>%
    mutate(`Hardy Score`= factor(DTHHRDY, levels = c(1,2,3,4,0)))

```

Ask questions below
```{r}
ggplot(final_dat, aes(x = SMTSISCH, y = G, color = as.factor(DTHVNT))) +
    geom_point(alpha = .5)

lm_dat <- filter(final_dat, DTHVNT != 99, !is.na(DTHVNT))

summary(lm(G ~ SMTSISCH + DTHVNT + AGE + SEX, data = lm_dat))


ggplot(final_dat, aes(x = SMTSISCH, y = E.2, color = as.factor(DTHVNT))) +
    geom_point(alpha = .8) +
    theme_classic() +
    scale_color_viridis(option = "A", discrete = T)

ggplot(final_dat, aes(x = SMRIN, y = E.2, color = as.factor(DTHHRDY))) +
    geom_point(alpha = .5)

for_plot <- filter(final_dat, !is.na(`Hardy Score`))

ggplot(for_plot, aes(x = SMTSISCH, y = E.2, color = `Hardy Score`)) +
    geom_point(alpha = .8) +
    theme_classic() +
    scale_color_viridis(option = "C", discrete = T,end = .93) +
    theme(text = element_text(size=15)) +
    labs(x = "Sample Ischemic Time",
         y = "Cluster E.2 Z-score")


ggplot(for_plot, aes(x = `Hardy Score`, y = E.2, color = `Hardy Score`)) +
    ggforce::geom_sina() +
    geom_boxplot(width=0.1) +
    theme_classic() +
    scale_color_viridis(option = "C", discrete = T,end = .93) +
    theme(text = element_text(size=15), legend.position = "top") +
    labs(x = "Hardy Score",
         y = "Cluster E.2 Z-score")

ggplot(final_dat, aes(x = SMTSISCH, y = E.1, color = as.factor(DTHHRDY))) +
    geom_point(alpha = .5)

ggplot(final_dat, aes(x = SMTSISCH, y = D, color = as.factor(DTHHRDY))) +
    geom_point(alpha = .5)



```
Finding overlaps between muscle genes and fast and slow twitch genes in our
dataset
```{r}
mouse_twitch_dat <- read.csv("data_in/mouse_skeletal_fast_slow_markers.csv")
mouse_twitch_dat <- mutate(mouse_twitch_dat,
                           twitch = ifelse(average_logFC_scRNA.seq > 0, "fast", "slow"),
                           human_gene = toupper(Gene))
all_clusters<- AllGeneClusterPuller("../../MARCC output/variance_genes_kendall/")
muscle_cluster <- all_clusters$muscle_skeletal

fast_twitch <- filter(mouse_twitch_dat, twitch == "fast")$human_gene
slow_twitch <- filter(mouse_twitch_dat, twitch == "slow")$human_gene

unlist(lapply(muscle_cluster, function(x){ sum(x %in% fast_twitch)}))
unlist(lapply(muscle_cluster, function(x){ sum(x %in% slow_twitch)}))

## No clear fast vs slow twitch, maybe E.2 has more fast twitch and A has more
## slow twitch but that's about it. Let me find out which genes come from which
## clusters though

fast_col <- t(as.data.frame(lapply(
    muscle_cluster, function(x){ paste(x[x %in% fast_twitch],collapse = "; ")})))
slow_col <- t(as.data.frame(lapply(
    muscle_cluster, function(x){ paste(x[x %in% slow_twitch],collapse = "; ")})))
twitch_comparison <- cbind(fast_col,slow_col)
colnames(twitch_comparison) <- c("fast_twitch", "slow_twitch")
```
Limit data to only ventillator individuals to test the "active before death" hypothesis

```{r}

vent_final <- final_dat %>%
    filter(DTHVNT == 1) %>%
    mutate(vent_group = case_when(
        DTHVNTD >= 24 & TRISCHD > 500 ~ "group_2",
        DTHVNTD <24 & TRISCHD > 500 ~ "group_1",
        TRUE ~ "group_3"
    ))



ggplot(vent_final, aes(DTHVNTD, TRISCHD, color = factor(vent_group))) +
    geom_point() +
    labs(title = "Time on Vent Compared to Ischemic Time",
         x = "Log10 Hours on Ventillator",
         y = "Ischemic Time") +
    scale_x_continuous(
        trans = "log10",
        breaks = c(.5, 1, 5, 10, 24, 100, 300, 1000, 10000)
    )


ggplot(vent_final, aes(x = SMTSISCH, y = E.2, color = as.factor(vent_group))) +
    geom_point(alpha = .5)

ggplot(final_dat, aes(x =  as.factor(DTHHRDY), y = E.2, color = as.factor(DTHHRDY))) +
    ggforce::geom_sina(alpha = .5)

ggplot(vent_final, aes(x =  as.factor(vent_group), y = E.2, color = as.factor(vent_group))) +
    ggforce::geom_sina(alpha = .5)



ggplot(final_dat, aes(x =  as.factor(MHNPHYS4W), y = E.2, color = as.factor(MHNPHYS4W), shape = as.factor(DTHHRDY))) +
    ggforce::geom_sina(alpha = .5)


```

Histology randomizer
```{r}
top_muscle <- head(arrange(final_dat, desc(E.2)), 5)
bot_muscle <- head(arrange(final_dat, (E.2)), 5)

pre_shuffle <- rbind(top_muscle, bot_muscle) %>% select(SAMPID) %>%
    mutate(Marc_notes = "")
set.seed(123)
shuffler <- sample(nrow(pre_shuffle))
shuffled <- pre_shuffle[shuffler,]
write.csv(shuffled,"data_out/blinded_muscle_samples.csv",row.names = F)

shuffled$SAMPID %in% top_muscle$SAMPID


marc_labels <- read.csv(file = "data_in/blinded_muscle_samples_marc_labeled.csv")

marc_labels <- left_join(marc_labels, final_dat)


ggplot(marc_labels, aes(Marc_notes, E.2)) +
    ggforce::geom_sina()

```

Fat atrophy checker

```{r}
final_dat <- final_dat %>%
    mutate(fat_mention = case_when(
    str_detect(SMPTHNTS, "fat") ~ TRUE,
    TRUE ~ FALSE),
    atrophy_mention = case_when(
        str_detect(SMPTHNTS, "atroph") ~ TRUE,
                   TRUE ~ FALSE)
    )



ggplot(final_dat, aes(x =  atrophy_mention, y = E.2, color =atrophy_mention)) +
    ggforce::geom_sina(alpha = .5)


ggplot(final_dat, aes(x =  fat_mention, y = E.2, color =fat_mention)) +
    ggforce::geom_sina(alpha = .5)


test_out <- CorrectionPlotter(final_dat, "E.2", c("AGE", "SEX"),"SMTSISCH")

test_out$plot + geom_point(aes(color = as.factor(DTHHRDY)))
```

Compare our results with people who have critical illness myopathy

```{r}
sig_CIM_genes <- data.table::fread("https://static-content.springer.com/esm/art%3A10.1186%2Fs13395-019-0194-1/MediaObjects/13395_2019_194_MOESM2_ESM.txt") %>%
    filter(FDR < 0.01)


sig_CIM_genes <- sig_CIM_genes %>%
    mutate(direction = ifelse(`FoldChange(ICU Patients vs. Control)` > 0,
                              "higher_in_CIM",
                              "lower_in_CIM"))

high_ICU_genes <- sig_CIM_genes$`Column ID`[sig_CIM_genes$direction =="higher_in_CIM"]
low_ICU_genes <- sig_CIM_genes$`Column ID`[sig_CIM_genes$direction =="lower_in_CIM"]


high_icu_col <- t(as.data.frame(lapply(
    muscle_cluster, function(x){ paste(x[x %in% high_ICU_genes],collapse = "; ")})))
low_icu_col <- t(as.data.frame(lapply(
    muscle_cluster, function(x){ paste(x[x %in% low_ICU_genes],collapse = "; ")})))
icu_comparison <- as.data.frame(cbind(high_icu_col,low_icu_col))
icu_comparison

TotalOverlapFunction <- function(x, icu_df){sum(x %in% icu_df)}
AllGenesInCluster <- function(x){length(x)}
overlap_high_icu <- lapply(muscle_cluster,  function(x, icu_df){sum(x %in% high_ICU_genes)})
overlap_low_icu <- lapply(muscle_cluster,  function(x, icu_df){sum(x %in% low_ICU_genes)})
total_genes_in_clust <- lapply(muscle_cluster, function(x){length(x)})

pre_bind <-as.data.frame(cbind(unlist(total_genes_in_clust),
                               unlist(overlap_high_icu),
                               unlist(overlap_low_icu)))

new_cim_table <- cbind(pre_bind,icu_comparison)
colnames(new_cim_table) <- c("N of genes", "N overlap with MIC upreg", "N overlap MIC downreg",
                             "Genes overlap MIC upreg", "Genes overlap MIC downreg")
#TNN1 downregulate in MIC according to paper, meaning lower levels are associated with MIC

write.csv(new_cim_table, "data_out/CIM_overlaps.csv")

## Make a 2 by 2 chi-sqare test
all_low_n_E2_genes <- data.frame(genes = unique(c(muscle_cluster$E.2,sig_CIM_genes$`Column ID`)))
all_low_n_E2_genes <- all_low_n_E2_genes %>%
    mutate(in_paper = genes %in% low_ICU_genes,
           in_cluster = genes %in% muscle_cluster$E.2)

table(all_low_n_E2_genes$in_paper, all_low_n_E2_genes$in_cluster)

## Marcs idea

all_low_n_all_genes <- data.frame(genes = unique(c(unlist(muscle_cluster),sig_CIM_genes$`Column ID`)))

all_low_n_all_genes <- all_low_n_all_genes %>%
    mutate(in_paper = case_when(
        genes %in% low_ICU_genes ~ "Low MIC gene",
        TRUE ~ "Not in low MIC"
        
    ),
           in_cluster = case_when(
               genes %in% muscle_cluster$E.2 ~ "In E.2",
               TRUE ~ "Not in E.2"
           ))

table(all_low_n_all_genes$in_paper, all_low_n_all_genes$in_cluster)

## Logic one: Of all cluster genes how many are in E.2
## of all down reg gene how many are in E2

## Another idea
sig_CIM_genes <- sig_CIM_genes %>%
    mutate(in_E2 = `Column ID` %in% muscle_cluster$E.2)
chisq_out <- chisq.test(sig_CIM_genes$in_E2, sig_CIM_genes$direction)

chisq_out$observed
chisq_out$expected


## Matts analysis

sig_CIM_genes <- mutate(sig_CIM_genes,
                        in_high_var = `Column ID` %in% unlist(muscle_cluster))

table(sig_CIM_genes$direction, sig_CIM_genes$in_high_var)

matt_fish_1 <- fisher.test(sig_CIM_genes$direction, sig_CIM_genes$in_high_var)

ind <- which(sig_CIM_genes$in_high_var)

chisq.test(sig_CIM_genes$direction, sig_CIM_genes$in_high_var)$observed
chisq.test(sig_CIM_genes$direction[ind], sig_CIM_genes$in_E2[ind])$observed
## Fish 2

table(sig_CIM_genes$direction[ind], sig_CIM_genes$in_E2[ind])

matt_fish_2 <- fisher.test(sig_CIM_genes$direction[ind], sig_CIM_genes$in_E2[ind])

## Are these eqtls?
sig_muscle_eqtls <- data.table::fread("data_in/Muscle_Skeletal.v8.egenes.txt/Muscle_Skeletal.v8.egenes.txt") %>%
    filter(qval < 0.05)

## Get muscle_cluster overlaps
overlapped_CIM_genes <- intersect(low_ICU_genes, muscle_cluster$E.2)


sig_muscle_eqtls_overlap <- sig_muscle_eqtls %>%
    filter(gene_name %in% overlapped_CIM_genes)
nrow(sig_muscle_eqtls_overlap)
e2_eqtls_overlap <- sig_muscle_eqtls %>%
    filter(gene_name %in% muscle_cluster$E.2)
nrow(e2_eqtls_overlap)

## Print out rsids for steph
write.table(e2_eqtls_overlap$rs_id_dbSNP151_GRCh38p7,
            "data_out/VCF_to_pull_for_muscle.txt",row.names = F,col.names = F,quote = F)

```

```{r}
ggplot(final_dat, aes(factor(DTHHRDY), J, color = factor(DTHHRDY))) +
    ggforce::geom_sina()
```

Test if eQTLS are effected by hardy score

Read in the various data forms
```{r}
require(vcfR)
## BED files of counts
eqtl_exprs <- data.table::fread("data_in/Muscle_Skeletal.v8.normalized_expression.bed",header = T)

## read in VCF
skeletal_vcf <- read.vcfR("data_in/GTEx_57_muscle_eQTLs.vcf")
skeletal_snp_dat <- as.data.frame(cbind(skeletal_vcf@fix, skeletal_vcf@gt))

## Read in eqtl covariates
skeletal_cov <- data.table::fread("data_in/Muscle_Skeletal.v8.covariates.txt") %>%
    column_to_rownames(var = "ID") %>%
    t() %>%
    as.data.frame() %>%
    rownames_to_column("SUBJID")

covariate_string <- paste0(colnames(skeletal_cov)[-1],
                           collapse = " + ")
```

Transform data to be used in analysis
Columns will be genes, rows will be individuals.
```{r}
## Filter genes to ones of interest.
rsids_avaialable <- filter(e2_eqtls_overlap, rs_id_dbSNP151_GRCh38p7 %in% skeletal_snp_dat$ID)
eqtl_exprs <- filter(eqtl_exprs, gene_id %in% rsids_avaialable$gene_id)
## get a gene table
gtab <- select(rsids_avaialable, gene_id, gene_name, ID = rs_id_dbSNP151_GRCh38p7) %>% unique()

## Create hardy dat
hardy_dat <- phen_dat %>%
    select(SUBJID, DTHHRDY) %>%
    mutate(DTHHRDY_factor = factor(DTHHRDY, levels = c("1", "2", "3", "4", "0")))

## Make an exprs matrix to loop through
loop_gene_tab <- eqtl_exprs %>% select(-`#chr`, -start, -end) %>% left_join(gtab, .)


```


Attempt the analysis with a for loop to make it easier and readable
```{r}

plot_list <- list()
steph_plots <- list()
analysis_out <- NULL
analysis_out_linear <- NULL
chi_table <- NULL
simple_output <- NULL
hold_off_genes <- c("CTC-338M12.1", "CTD-2083E4.7", "CTD-3064M3.7")

for (rsid_ind in seq(nrow(rsids_avaialable))) {
    cur_rsid <- rsids_avaialable$rs_id_dbSNP151_GRCh38p7[rsid_ind]
    cur_gene_id <- rsids_avaialable$gene_id[rsid_ind]
    cur_gene_name <- rsids_avaialable$gene_name[rsid_ind]

    gene_col <- filter(loop_gene_tab, ID == cur_rsid) %>% 
        select(-gene_id, -gene_name, -ID) %>%
        t() %>%
        as.data.frame() %>%
        rownames_to_column("SUBJID")
    colnames(gene_col)[2] <- cur_gene_id
    snp_col <- filter(skeletal_snp_dat, ID == cur_rsid) %>%
        select(starts_with("GTEX")) %>%
        t() %>%
        as.data.frame() %>%
        rownames_to_column("SUBJID")
    colnames(snp_col)[2] <- cur_rsid
    analysis_dat <- left_join(snp_col, gene_col) %>%
        left_join(., hardy_dat) %>% drop_na() %>%
        left_join(.,skeletal_cov)
    
    
    ## Chisq analysis
    chi_res <- chisq.test(analysis_dat[,2], analysis_dat$DTHHRDY_factor)
    chi_sig <- chi_res$p.value <0.05
    new_chi_row <- c(broom::tidy(chi_res), cur_rsid, cur_gene_name)
    chi_table <- rbind(chi_table, new_chi_row)
    ##Plotting
    
    cur_plot_simple <- ggplot(analysis_dat, aes_string(x = cur_rsid, y = cur_gene_id)) +
        ggforce::geom_sina() +
        labs(y= paste0(cur_gene_name))
    
    cur_plot_hardy <- ggplot(analysis_dat, aes_string(x = cur_rsid, y = cur_gene_id,
                                    color = "DTHHRDY_factor")) +
        ggforce::geom_sina() +
        labs(y= paste0(cur_gene_id),
             subtitle = paste0("Sig chi: ",chi_sig))
    
    formula_string <- paste0(cur_gene_id, " ~ ", cur_rsid, " + DTHHRDY_factor", "+", covariate_string)

    
    
    ## Factor model
    temp_model <- lm(as.formula(formula_string), data = analysis_dat)
   
    new_rows <- cbind(broom::tidy(summary(temp_model)), cur_gene_name)
    analysis_out <- rbind(analysis_out, new_rows)
    
    
    
    
    ## Linear model
    analysis_dat_linear <- analysis_dat
    analysis_dat_linear[,2] <- case_when(
        analysis_dat_linear[,2] == "0/0" ~ 0,
        analysis_dat_linear[,2] == "0/1" ~ 1,
        analysis_dat_linear[,2] == "1/1" ~ 2
    )
    
    linear_model <- lm(as.formula(formula_string), data = analysis_dat_linear)
    ## Steph models
    ## exprs ~ SNP
    simple_model <- lm(as.formula(paste0(cur_gene_id, " ~ ", cur_rsid)),
                       data = analysis_dat_linear)
    
    ## simple models out
    simple_rows <- cbind(broom::tidy(simple_model),
                         cur_gene_name, cur_rsid)
    
    simple_output <- rbind(simple_output, simple_rows)
    
    ## Steph comparison anova
    anova_out <- anova(simple_model,linear_model)
    anova_sig <- broom::tidy(anova_out)$p.value[2] < 0.05
    
   
    new_linear <- cbind(broom::tidy(summary(linear_model)), cur_gene_name, anova_sig)
    analysis_out_linear <- rbind(analysis_out_linear, new_linear)
    
    
    analysis_dat_linear <- mutate(analysis_dat_linear,
        DTHHRDY_linear = ifelse(DTHHRDY == 0, 5, DTHHRDY)
    )
    ## steph plot
    steph_plot <- ggplot(analysis_dat_linear, aes_string(x = cur_rsid, y = "DTHHRDY_linear")) +
        geom_jitter(aes(color = DTHHRDY_factor)) +
        geom_smooth(method = 'lm')
   
    steph_plots[[cur_rsid]] <- steph_plot
    
    ## Combine the plots
    comb_plot <- cur_plot_simple / cur_plot_hardy
    
    plot_list[[cur_rsid]] <- comb_plot
}

pdf("data_out/eqtl_vs_hardy.pdf", width = 8, height = 10)
for (plot_ind in seq_along(plot_list)) {
    print(plot_list[[plot_ind]])
}
dev.off()
```
Linear Analysis
```{r}
adj_linear <- analysis_out_linear %>% filter((startsWith(term, "rs"))) %>%
    mutate(p_adj = p.adjust(p.value,method = "fdr"),
           sig_adj = p_adj < 0.05,
           gws_sig = p.value < 5e-8)

table(adj_linear$sig_adj)
table(adj_linear$gws_sig)

non_sig <- filter(adj_linear, sig_adj == F)

```


Redo above analysis with Z-score
```{r}
z_score_snps <- skeletal_snp_dat %>%
    select(ID, starts_with("GTEX")) %>%
    t() %>%
    as.data.frame() 

colnames(z_score_snps) <- z_score_snps[1,]
z_score_snps_final <- z_score_snps[-1,] %>%
    rownames_to_column("SUBJID") %>%
    left_join(final_dat,.) %>%
    mutate(DTHHRDY_factor = factor(DTHHRDY, levels = c("1","2","3","4","0")))


zscore_eqtl_plots <- list()
for (rsid_ind in seq(nrow(skeletal_snp_dat))) {
    cur_rsid <- skeletal_snp_dat$ID[rsid_ind]
    na_ind <- !(is.na(z_score_snps_final[,cur_rsid]) | is.na(z_score_snps_final$DTHHRDY_factor))
    plot_dat <- z_score_snps_final[na_ind,]
    classic_plot <- ggplot(plot_dat, aes_string(cur_rsid, "E.2")) +
        ggforce::geom_sina()
    hardy_plot <- ggplot(plot_dat, aes_string(cur_rsid, "E.2",
                                                          color = "DTHHRDY_factor")) +
        ggforce::geom_sina()
    zscore_plot <- classic_plot/hardy_plot
    zscore_eqtl_plots[[cur_rsid]] <- zscore_plot
    
    
}

```


Do covariates correlate with Hardy?

```{r}
hardy_cov_test <- left_join(hardy_dat, skeletal_cov) %>%
    mutate(DTHHRDY_linear = ifelse(DTHHRDY == 0, 5, DTHHRDY))

hardy_inferred <- select(hardy_cov_test, DTHHRDY_linear, starts_with("Inferred")) %>%
    drop_na()

cor(hardy_inferred)

(cor.test(hardy_inferred$DTHHRDY_linear, hardy_inferred$InferredCov1))

ggplot(hardy_cov_test, aes(DTHHRDY_factor, InferredCov1)) +
    ggforce::geom_sina()
```


```{r}
lm_dat <- mutate(lm_dat,
    DTHHRDY_factor = factor(DTHHRDY, c(1,2,3,4,0)),
    in_hospital_death = ifelse(DTHHRDY %in% c(1,2), FALSE, TRUE)
)

model_1 <- lm(E.2 ~ DTHHRDY_factor + AGE + SEX+ SMTSISCH, data = lm_dat)
model_2 <- lm(E.2 ~ in_hospital_death + AGE + SEX + SMTSISCH, data = lm_dat)

summary(model_1)
summary(model_2)

AIC(model_1)
AIC(model_2)
tidy_model_1 <- broom::tidy(model_1)
write.csv(tidy_model_1, "data_out/tidy_model_for_talk.csv")


only_hardy <- lm(E.2 ~ DTHHRDY_factor + AGE + SEX, data = lm_dat)
only_isch <- lm(E.2 ~ SMTSISCH + AGE + SEX, data = lm_dat)

```

